src/btree/bt_cursor.c

```diff
@@ -1964,6 +1964,16 @@ __wt_btcur_range_truncate(WT_CURSOR_BTREE *start, WT_CURSOR_BTREE *stop)
 
     WT_STAT_DATA_INCR(session, cursor_truncate);
 
+    /*
+     * All historical versions must be removed when a key is updated with no timestamp, but that
+     * isn't possible in fast truncate operations. Disallow fast truncate in transactions configured
+     * to commit without a timestamp (excluding logged objects as timestamps will be ignored).
+     */
+    if (!F_ISSET(btree, WT_BTREE_LOGGED) && F_ISSET(session->txn, WT_TXN_TS_NOT_SET))
```

Comment:
This seems like a reasonable starting point, but we might need to figure out a way to move this check to later in the transaction lifecycle when MongoDB starts using truncate on timestamped tables. They often don't know which timestamp an operation will use until after the operation is completed. Could you extend the message to be a bit more welcoming of a conversation? Something like: "...without a timestamp. If you have a use case the encounters this, but expects a timestamp to be assigned before commit, please reach out to the WiredTiger team to discuss."
It's kind of janky - but I wouldn't want someone to give up on using timestamped truncates without asking..

Reply:
@agorrod, I think I might be missing your point, but since the only way to get a timestamp=0 on commit is to set the `no-timestamp` flag during `begin_transaction`, I think we can know at this point whether or not it's possible for the fast-truncate to ever have a zero timestamp at commit.

I've enhanced the error message as you suggested in 98c3a96, it now says:

> "truncate operations may not yet be included in transactions that can commit without a timestamp. If your use case encounters this error, please reach out to the WiredTiger team"